pytest/defs.bzl

```diff
@@ -0,0 +1,65 @@
+"""Use pytest to run tests, using a wrapper script to interface with Bazel.
+
+Example:
+
+```starlark
+load("@aspect_rules_py//pytest:defs.bzl", "pytest_test")
+
+pytest_test(
+    name = "test_w_pytest",
+    size = "small",
+    srcs = ["test.py"],
+)
+```
+
+By default, `@pip//pytest` is added to `deps`.
+If sharding is used (when `shard_count > 1`) then `@pip//pytest_shard` is also added.
+To instead provide explicit deps for the pytest library, set `pytest_deps`:
+
+```starlark
+pytest_test(
+    name = "test_w_my_pytest",
+    shard_count = 2,
+    srcs = ["test.py"],
+    pytest_deps = [requirement("pytest"), requirement("pytest-shard"), ...],
+)
+```
+"""
+
+load("//py:defs.bzl", "py_test")
+
+def pytest_test(name, srcs, deps = [], args = [], pytest_deps = None, pip_repo = "pip", **kwargs):
+    """
+    Wrapper macro for `py_test` which supports pytest.
+
+    Args:
+      name: A unique name for this target.
+      srcs: Python source files.
+      deps: Dependencies, typically `py_library`.
+      args: Additional command-line arguments to pytest.
+        See https://docs.pytest.org/en/latest/how-to/usage.html
+      pytest_deps: Labels of the pytest tool and other packages it may import.
+      pip_repo: Name of the external repository where Python packages are installed.
+        It's typically created by `pip.parse`.
+        This attribute is used only when `pytest_deps` is unset.
+      **kwargs: Additional named parameters to py_test.
+    """
+    shim_label = Label("//pytest:pytest_shim.py")
+
+    if pytest_deps == None:
+        pytest_deps = ["@{}//pytest".format(pip_repo)]
+        if kwargs.get("shard_count", 1) > 1:
+            pytest_deps.append("@{}//pytest_shard".format(pip_repo))
+
+    py_test(
+        name = name,
+        srcs = [
+            shim_label,
+        ] + srcs,
+        main = shim_label,
+        args = [
+            "--capture=no",
```

Comment:
What's the reason to have this defined here, rather than in the shim?

Reply:
I don't know, I didn't write this code. Maybe Casey is willing to be the PR author.

Reply:
This absolutely can be in the shim. In the very beginning, the shim was a trivial wrapper around `pytest.main` and adding default arguments here was less code than in the shim. As more pytest features were hooked up to native Bazel features, the shim gathered more argument parsing logic and I simply missed calling this out in the pull request that first began adding default arguments in the shim. I see no reason this shouldn't be part of `pytest_args = ["--ignore=external"]` in the shim.